text/0000-partial-turbofish.md

```diff
@@ -0,0 +1,314 @@
+rustc will complain that "the trait bound `Vec<u32>: FromIterator<i32>` is not
+satisfied". Since the only way to satisfy `Vec<T>: FromIterator<U>` is with
+`impl<T> FromIterator<T> for Vec<T>`, unifying `T` with `U`, subsituted to `u32`
+and `i32` fails.
+
+Starting from `turbo::<Vec<u32>, _>(1, 1)`, the compiler can first see
+that `T = Vec<u32>`. It can then see that the only way for
+`Vec<u32>: FromIterator<U>` is if `U = u32`. If `turbo::<Vec<_>, _>(true, 1)`
+is used, then rustc can start from `U = bool` and so it may infer from the other
+direction.
+
+In either case, the argument to the second type parameter is known which is why
+rustc allows `_` instead of a concrete type.
+
+This RFC proposes that suffixes to a "turbofish" which consist entirely of
+`, _`s be omittable.
+
+The following:
+
+```rust
+fn main() {
+    let vec = turbo::<Vec<u32>, _>(1, 1);
+    let vec = Turbo::<Vec<u32>, _>::new(1, 1);
+}
+```
+
+thus becomes:
+
+```rust
+fn main() {
+    let vec = turbo::<Vec<u32>>(1, 1);
+    let vec = Turbo::<Vec<u32>>::new(1, 1);
+}
+```
+
+This concept is named *partial turbofish* since it is partially applying
+type arguments to type parameters and letting the compiler infer the rest.
+When `, _` is present in a turbofish, it is instead referred to as *turbosword*.
+
+It is still an error to apply too many type arguments in a turbofish
+as in: `fn foo() {} foo::<i32>()`, or to partially apply types when the
+compiler can not infer the omitted types. When the latter happens, rustc will
+emit error indicating that there are more type parameters not applied that
+could not be inferred. A sample error message is: `type annotations needed`
+in addition to:
+
+```
+Can not infer concrete type(s) of omitted type parameter(s) `X, Y, ..`
+in call to `fn turbo` with parameters: `<Concrete1, X, Y, ...>`.
+```
+
+Some developers may end up trying to use this feature and be surprised when it
+did not work. For those who this does not apply to, the documentation should
+explain this feature when it explains the turbofish syntax.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC allows the user to only supply the smallest prefix of
+concrete types that rustc requires to infer the type of the other parameters in
+the suffix. The suffix may therefore be omitted. The prefix may be overspecified
+and the omitted suffix may be shorter than the longest possible.
+
+Formally: Assume a function (free, inherent, trait, ...)
+`fn turbo< $($tparam: ident),* )>(..) -> R` or a generic type (struct/enum/..)
+`Turbo` with type parameters `$($tparam: ident),*`. If while calling
+`turbo::< $($tconcrete: ty),* >(..)` or equivalently
+`Turbo::< $($tconcrete: ty),* >::fun(..)`, for a list of concrete types
+`$($tconcrete: ty),*`, a suffix of `$($tconcrete: ty),*` is `$( _ ),*` and
+the type checking passes, then that suffix may be omitted.
```

Comment:
I usually seen these called "path arguments in *value context*", as I understand this RFC touches only them.

- Value context in (function) bodies is `foo::<_>(a, b)`/`Vec::<_>::foo`/etc - 1) `_` can be inferred, 2) missing arguments are replaced with `_`s (before this RFC) if no explicit arguments are provided  or (after this RFC) always, and 3) how exactly `_`s are inferred is an orthogonal question, the inference can be enhanced with fallback based on default type parameters.
- Type context in (function) bodies is `let x: Vec<_>`/`x.collect::<Vec<_>>`/etc - 1) `_` can be inferred, 2) missing arguments are *not* replaced with `_`s but can be replaced by default type parameters, `_`s have to be written manually, and 3) how exactly `_`s are inferred is an orthogonal question, the inference can be enhanced with fallback based on default type parameters.
- Type context in signatures is `fn f(arg: Vec<_>)`/`const C: Vec<_>` - 1) `_` cannot be inferred and is prohibited, 2) missing arguments are *not* replaced with `_`s but can be replaced by default type parameters, and 3) `_`s cannot be inferred as previously mentioned.

Reply:
It would be interesting to apply the idea from this RFC (replacing missing arguments with `_`s) to the second case as well. Imagine how nice would be to write
```rust
x.collect::<Vec>()
let y: Vec = x.collect();
```
but that would be more challenging from backward-compatibility point of view, unfortunately.

Reply:
FWIW, I support alway replacing missing type arguments with `_`s for the case 1, as this RFC suggests.
This is certainly an ergonomic win, it's future-compatible with fallback-related inference changes and I've never seen this causing problems in C++.

Reply:
I think @withoutboats was opposed to allowing:

```rust
x.collect::<Vec>()
let y: Vec = x.collect();
```

There could be some problems relating to future HKT proposals.
We can certainly revisit this with an RFC at some point, but I prefer to only allow this in turbofish at this point in time.